velox/dwio/parquet/reader/DeltaByteArrayDecoder.h

```diff
@@ -0,0 +1,159 @@
+ */
+
+#pragma once
+
+#include "velox/common/base/BitUtil.h"
+#include "velox/dwio/parquet/reader/DeltaBpDecoder.h"
+#include "velox/dwio/parquet/reader/DeltaLengthByteArrayDecoder.h"
+
+namespace facebook::velox::parquet {
+
+using namespace velox::memory;
+
+// DeltaByteArrayDecoder is adapted from Apache Arrow:
+// https://github.com/apache/arrow/blob/apache-arrow-15.0.0/cpp/src/parquet/encoding.cc#L3301-L3545
+class DeltaByteArrayDecoder {
+ public:
+  explicit DeltaByteArrayDecoder(const char* start) {
+    prefixLenDecoder_ = std::make_unique<DeltaBpDecoder>(start);
+    int64_t numPrefix = prefixLenDecoder_->validValuesCount();
+    bufferedPrefixLength_.resize(numPrefix);
+    prefixLenDecoder_->readValues<uint32_t>(bufferedPrefixLength_, numPrefix);
+    prefixLenOffset_ = 0;
+    numValidValues_ = numPrefix;
+
+    suffixDecoder_ = std::make_unique<DeltaLengthByteArrayDecoder>(
+        prefixLenDecoder_->bufferStart());
+    lastValue_.clear();
+  }
+
+  void skip(uint64_t numValues) {
+    skip<false>(numValues, 0, nullptr);
+  }
+
+  template <bool hasNulls>
+  inline void skip(int32_t numValues, int32_t current, const uint64_t* nulls) {
+    if (hasNulls) {
+      numValues = bits::countNonNulls(nulls, current, current + numValues);
+    }
+    for (int32_t i = 0; i < numValues; ++i) {
+      readString();
+    }
+  }
+
+  template <bool hasNulls, typename Visitor>
+  void readWithVisitor(const uint64_t* nulls, Visitor visitor) {
+    int32_t current = visitor.start();
+    skip<hasNulls>(current, 0, nulls);
+    int32_t toSkip;
+    bool atEnd = false;
+    const bool allowNulls = hasNulls && visitor.allowNulls();
+    for (;;) {
+      if (hasNulls && allowNulls && bits::isBitNull(nulls, current)) {
+        toSkip = visitor.processNull(atEnd);
+      } else {
+        if (hasNulls && !allowNulls) {
+          toSkip = visitor.checkAndSkipNulls(nulls, current, atEnd);
+          if (!Visitor::dense) {
+            skip<false>(toSkip, current, nullptr);
+          }
+          if (atEnd) {
+            return;
+          }
+        }
+
+        // We are at a non-null value on a row to visit.
+        toSkip = visitor.process(readString(), atEnd);
+      }
+      ++current;
+      if (toSkip) {
+        skip<hasNulls>(toSkip, current, nulls);
+        current += toSkip;
+      }
+      if (atEnd) {
+        return;
+      }
+    }
+  }
+
+  folly::StringPiece readString() {
+    auto suffix = suffixDecoder_->readString();
+    bool isFirstRun = (prefixLenOffset_ == 0);
+    const int64_t prefixLength = bufferedPrefixLength_[prefixLenOffset_++];
+
+    VELOX_CHECK_GE(
+        prefixLength, 0, "negative prefix length in DELTA_BYTE_ARRAY");
+
+    std::string_view prefix{lastValue_};
+    std::string tempString;
+    tempString.resize(prefixLength + suffix.size());
+    buildBuffer(isFirstRun, prefixLength, suffix, &prefix, tempString);
+
+    numValidValues_--;
+    lastValue_ = std::string{prefix};
```

Comment:
Not sure why we need to do a copy here, just declare `prefix` as `StringPiece` directly and return it

Reply:
I have made modifications to this part using `folly::StringPiece`.

Reply:
My point is that you don't need to keep 2 buffers `lastValue_` and `dataBuffer_`.  Just keep a single buffer, and keep the first `prefixLength` bytes every time you overwrite it.  This way you can avoid copy of the whole buffer here.

Reply:
You can remove this line now